### PR TITLE
Compact printing.

### PR DESCRIPTION
**Pull request by mbebenita**
This makes it possible to print a wast file on a single line with minimal whitespace. The motivation for this patch was that I wanted to be able to print AST subtrees on a single line for the purpose of displaying metrics. Then I figured I might as well minify the entire module.

I can't figure out an easy way to add arguments to passes. (It would be nice to be able to say `--print --compact` or something like that.) So at the moment there is no way to actually enable this from the command line. (I didn't want to add an entire new PrintCompact.cpp file just to do this).

You can however try it out if you use `--metrics`, this will print AST subtrees on a single line:

```
Patterns
 15: (set_local $(i32.shr_u(get_local $)(none.const ?)))
 15: (set_local $(i32.and(i32.shr_u(i32.add(get_local $)(none.const ?))(none.const ?))(none.const ?)))
 15: (set_local $(i32.and(i32.shr_u(get_local $)(none.const ?))(none.const ?)))
```

Notice the weird `$` names, I added an option to the Print pass to erase names and values, this is useful when looking for patterns.

```
 2155: (get_local $)
 255: (set_local $(get_local $))
 131: (set_local $(i32.add(get_local $)(none.const ?)))
```

### CODE REVIEW

**Comment by mbebenita** [comment]
If we add options to passes then I can also add a test case that ensures the compact output is parsed back the same way. I tried this locally and it works, but it would be nice to add this to the testing suite.

**Comment by mbebenita** [comment]
Metrics from banana bread: https://pastebin.mozilla.org/8860409

You can easily add your own by following this example: https://github.com/WebAssembly/binaryen/pull/204/files#diff-34ae3913118ed6b9c570f9c5d6b997b4R49

File: src/wasm-s-parser.h [hunk 1]
```diff
@@ -205,7 +205,7 @@ class SExpressionParser {
       input++;
       return allocator.alloc<Element>()->setString(IString(str.c_str(), false), dollared);
     }
-    while (input[0] && !isspace(input[0]) && input[0] != ')') input++;
+    while (input[0] && !isspace(input[0]) && input[0] != ')' && input[0] != '(') input++;
```
Comment (kripken) [inline src/wasm-s-parser.h]: what is this for?

**Comment by kripken** [comment]
It might be nice to add options for passes, but I don't have a good idea of how to do it. Meanwhile, though, look at the bottom of `Print.cpp` where we have `static RegisterPass<Printer> registerPass("print", "print in s-expression format");` - we can register multiple passes in a single file. There could be `print-minified` for example.

File: src/passes/Metrics.cpp [hunk 2]
```diff
@@ -35,6 +39,21 @@ struct Metrics : public WalkerPass<WasmWalker<Metrics>> {
...
+  }
+  void visitConst(Const *curr) {
+    patterns[getExpressionText(curr)]++;
+  }
+  void visitGetLocal(GetLocal *curr) {
+    patterns[getExpressionText(curr, true)]++;
+  }
+  void visitSetLocal(SetLocal *curr) {
+    patterns[getExpressionText(curr, true)]++;
+  }
```
Comment (kripken) [inline src/passes/Metrics.cpp]: what is special about const, getlocal and setlocal?

File: src/wasm-s-parser.h (hunk 1 above)
Comment (mbebenita) [inline src/wasm-s-parser.h]: If I remember correctly, `(module(memory ...)` doesn't parse unless you stop parsing `memory` when you see the `(`.

File: src/passes/Metrics.cpp (hunk 2 above)
Comment (mbebenita) [inline src/passes/Metrics.cpp]: Nothing really, just examples. I should just comment it out.

**Comment by mbebenita** [comment]
K, so this patch is a bit sloppy. I should remove the name erasing stuff as it's not that important. I was just experimenting with stuff.

File: src/wasm-s-parser.h (hunk 1 above)
Comment (kripken) [inline src/wasm-s-parser.h]: would be good to split that out into a bugfix with a testcase, then. but surely this would already be tested?

**Comment by mbebenita** [comment]
I removed all the metric stuff and just left the minification code in there.

**Comment by mbebenita** [comment]
@kripken I, split this PR in two PRs, #210 applies the bug fix, and this adds whitespace minification.

File: src/pass.h [hunk 3]
```diff
@@ -165,6 +166,15 @@ class Printer : public Pass {
   void run(PassRunner* runner, Module* module) override;
 };
 
+// Prints out a minified module
```
Comment (kripken) [inline src/pass.h]: Do we need this here? Printer is here so other places can use it, e.g. for debugging. But if this is only used by the user specifying it as a commandline pass, it doesn't need to be here.

Comment (mbebenita) [inline src/pass.h]: Not really. I can move it out.

**Pull request merged**